src/libopensc/card-idprime.c

```diff
@@ -380,11 +567,13 @@ static int idprime_fill_prkey_info(list_t *list, idprime_object_t **entry, sc_pk
 	/* Do not specify the length -- it will be read from the FCI */
 	prkey_info->path.count = -1;
 
-	/* TODO figure out the IDs as the original driver? */
-	prkey_info->id.value[0] = ((*entry)->fd >> 8) & 0xff;
-	prkey_info->id.value[1] = (*entry)->fd & 0xff;
-	prkey_info->id.len = 2;
+	memcpy(prkey_info->id.value, (*entry)->key_id, 8);
+	prkey_info->id.len = 8;
```

Comment:
My IDPrime does not have the key id property so it falls back to the counter, but causes 7 bytes zero padding, which does not look great. Can the fallback use the length of 2 again? Probably by storing the length of the key_id alongside it.

Reply:
should be solved with  55f3c4d

Reply:
I'm not sure the functions I have added for extracting the key info are good.

Could you toggle `#if 0` here with your card and see if it logs correct fields?
https://github.com/OpenSC/OpenSC/blob/55f3c4d36ad9ae64d57e099e91b9cdbeb78a44f7/src/libopensc/card-idprime.c#L172

They work in my case but I'm skipping some bytes here:
https://github.com/OpenSC/OpenSC/blob/55f3c4d36ad9ae64d57e099e91b9cdbeb78a44f7/src/libopensc/card-idprime.c#L242
The `if()` in that line that decides how many bytes to skip and is likely wrong.
I had '1' there on CA certificates and '2' on personal certs.

If you don't have time for that would appreciate if you could dump the pubk[sx]cNNp11 files from the card and share them somehow (maybe just ones for CA keys so you don't share your personal info) and will try to improve parsing more.

Reply:
I suppose you're getting correct key labels now?
Those seem to be at fixed position from pubk file start for me.
The key_id is immediately following certificate string tags (C/O/CN/L/...) - problem is those tags do not start at constant offset from file start :(